src/commands/git-metadata/gitdb.ts

```diff
@@ -0,0 +1,297 @@
+    throw err
+  }
+
+  let packfiles
+  try {
+    packfiles = generatePackFilesForCommits(log, objectsToUpload)
+    log.debug(`${packfiles.length} packfiles generated.`)
+  } catch (err) {
+    log.warn(`Failed generating packfiles: ${err}`)
+    throw err
+  }
+
+  if (dryRun) {
+    log.debug(`Dry-run enabled, not uploading anything.`)
+
+    return
+  }
+  log.debug(`Uploading packfiles...`)
+  try {
+    await uploadPackfiles(log, request, repoURL, latestCommits[0], packfiles)
+    log.debug('Successfully uploaded packfiles.')
+  } catch (err) {
+    log.warn(`Failed to upload packfiles: ${err}`)
+    throw err
+  }
+}
+
+const getRepoURL = gitRemote
+
+const getLatestLocalCommits = async (git: simpleGit.SimpleGit) => {
+  const logResult = await git.log(['-n 1000', '--since="1 month ago"'])
+
+  return logResult.all.map((c) => c.hash)
+}
+
+const unshallowRepositoryWhenNeeded = async (log: Logger, git: simpleGit.SimpleGit) => {
+  const isShallow = (await git.revparse('--is-shallow-repository')) === 'true'
+  if (isShallow) {
+    log.info('[unshallow] Git repository is a shallow clone., unshallowing it...')
+    log.info('[unshallow] Setting remote.origin.partialclonefilter to "blob:none" to avoid fetching file content')
+    await git.addConfig('remote.origin.partialclonefilter', 'blob:none')
+    log.info('[unshallow] Running git fetch --shallow-since="1 month ago" --update-shallow --refetch')
+    await git.fetch(['--shallow-since="1 month ago"', '--update-shallow', '--refetch'])
+    log.info('[unshallow] Fetch completed.')
+  }
+}
+
+// getKnownCommits asks the backend which of the given commits are already known
+const getKnownCommits = async (log: Logger, request: RequestBuilder, repoURL: string, latestCommits: string[]) => {
+  interface SearchCommitResponse {
+    data: Commit[]
+  }
+
+  interface Commit {
+    type: string
+    id: string
+  }
+
+  const localCommitData = JSON.stringify({
+    meta: {
+      repository_url: repoURL,
+    },
+    data: latestCommits.map((commit) => ({
+      id: commit,
+      type: 'commit',
+    })),
+  })
+  const response = await runRequest(log, 'search_commits', () =>
+    request({
+      url: '/api/v2/git/repository/search_commits',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      data: localCommitData,
+      method: 'POST',
+      timeout: API_TIMEOUT,
+    })
+  )
+  const commits = response.data as SearchCommitResponse
+  if (!commits || commits.data === undefined) {
+    throw new Error(`Invalid API response: ${response}`)
+  }
+
+  return commits.data.map((c) => {
+    if (c.type !== 'commit' || c.id === undefined) {
+      throw new Error('Invalid commit type response')
+    }
+
+    return sanitizeCommit(c.id)
+  })
+}
+
+const sanitizeCommit = (sha: string) => {
+  const isValidSha = (s: string) => /[0-9a-f]{40}/.test(s)
+
+  const sanitizedCommit = sha.replace(/[^0-9a-f]+/g, '')
+  if (sanitizedCommit !== sha) {
```

Comment:
I took that from the dd-trace-js implementation, but looking at the code I don't think there is any good reason.
For simplification I'll do two changes:
- replace the regex `/[0-9a-f]{40}/` by `/^[0-9a-f]{40}$/` to test exact match
- remove the sanitization part and only test against the regex above

cc @juan-fernandez fyi in case I'm missing something

Reply:
yeah good catch. I don't remember why I did it like this, but it's wrong 😅 . Not only because of the unnecessary replacement and test, but because the regex should indeed be `/^[0-9a-f]{40}$/`. I'll update it in `dd-trace-js`, thanks

Reply:
ok I now remember why I had this 😅 . It's because of this: 
https://github.com/DataDog/dd-trace-js/pull/2876/checks?check_run_id=11845280052 

```
Uncontrolled command line
Critical
This command line depends on a user provided value
```

It's kind of silly because if it's anything other than a SHA it will not go through, but by sanitizing the alert goes away.

Reply:
so I'll keep the sanitize in `dd-trace-js` but I'll update the regex anyway 👍